File: 3143-longest-unequal-adjacent-groups-subsequence-i/3143-longest-unequal-adjacent-groups-subsequence-i.java
class Solution {
    public List<String> getLongestSubsequence(String[] words, int[] groups) {

        Integer prev_ele = null;
        int n = words.length;
        List<String> res = new ArrayList<>();
        if (n==0) return res;

        for (int i=0; i<words.length; i++){
            if (prev_ele == null || groups[i] != prev_ele) {
                res.add(words[i]);
                prev_ele = groups[i];
            }
        }

        return res;
        
    }
}

File: 3143-longest-unequal-adjacent-groups-subsequence-i/3143-longest-unequal-adjacent-groups-subsequence-i.py
from typing import List

class Solution:
    def getLongestSubsequence(self, words: List[str], groups: List[int]) -> List[str]:
        n = len(groups)
        if n == 0:
            return []
        
        # result list and last-picked group
        res = []
        prev_group = None
        
        # scan through exactly once, picking on every flip
        for i in range(n):
            if prev_group is None or groups[i] != prev_group:
                res.append(words[i])
                prev_group = groups[i]
        
        return res
